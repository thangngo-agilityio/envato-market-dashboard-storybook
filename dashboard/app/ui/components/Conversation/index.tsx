import { memo } from 'react';
import { Box, Heading, Flex } from '@chakra-ui/react';

// Components
import { ChatMember, ListMessages } from '..';

// Interface
import { TMessages } from '@/lib/interfaces';

// Interfaces
import { MessageType } from '@/lib/interfaces/messages';

export type Props = {
  adminUid?: string;
  filteredMessages?: MessageType[];
  nameUser: string;
  avatarUser: string;
  messages: TMessages[];
};

const Conversation = ({ nameUser, avatarUser, messages, adminUid }: Props) => {
  const defaultName = nameUser;

  return (
    <Box w="full" borderRadius="lg">
      <Flex
        direction="row"
        justifyContent="space-between"
        padding={{ base: '8px', lg: '24px 26px' }}
        bg="background.body.septenary"
      >
        <Heading
          as="h3"
          fontWeight="semibold"
          color="text.primary"
          fontSize="2xl"
          textTransform="capitalize"
        >
          <ChatMember
            avatar={avatarUser}
            name={defaultName}
            statusColor="online"
          />
        </Heading>
      </Flex>

      <ListMessages
        avatarUser={avatarUser}
        messages={messages}
        adminUid={adminUid}
        nameUser={nameUser}
      />
    </Box>
  );
};

const ConversationMemorized = memo(Conversation);

export default ConversationMemorized;
