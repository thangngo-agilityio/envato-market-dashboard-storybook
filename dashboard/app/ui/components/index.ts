export { default as InputField } from './common/InputField';
export { default as FallbackErrorBoundary } from './FallbackErrorBoundary';
export { default as Navigation } from './common/Navigation';
export { default as Menu } from './common/Menu';
export { default as Logo } from './common/Logo';
export { default as ExpandSidebar } from './ExpandSidebar';
export { default as MiniSidebar } from './MiniSidebar';
export * from './Icons';
export * from './Images';
export { default as Indicator } from './Indicator';
export { default as Loading } from './common/Loading';
export { default as Avatar } from './common/Avatar';
export { default as Benefit } from './common/Benefit';
export { default as Divider } from './common/Divider';
export { default as IconButton } from './common/IconButton';
export { default as Dropdown } from './common/Dropdown';
export { default as Select } from './common/Select';
export { default as Pagination } from './common/Pagination';
export { default as Button } from './common/Button';
export { default as Table } from './common/Table';
export { default as CustomerNameCell } from './common/Table/Body/CustomerNameCell';
export { default as ProductNameCell } from './common/Table/Body/ProductNameCell';
export { default as ActionCell } from './common/Table/Body/ActionCell';
export { default as UserInfoCell } from './UsersTable/Cell/UserInfoCell';
export { default as StatusCell } from './UsersTable/Cell/StatusCell';
export { default as HeadCell } from './common/Table/HeadCell';
export { default as Selector } from './TransactionTable/Selector';
export { default as TransactionTable } from './TransactionTable';
export { default as SearchBar } from './common/SearchBar';
export { default as SwitchTheme } from './SwitchTheme';
export { default as CardPayment } from './CardPayment';
export { default as BoxChat } from './BoxChat';
export { default as TotalStatisticCard } from './TotalStatisticCard';
export { default as Fetching } from './Skeleton/Fetching';
export { default as TotalStatisticListSkeleton } from './Skeleton/TotalStatisticListSkeleton';
export { default as TotalStatisticList } from './TotalStatisticList';
export { default as RevenueFlow } from './RevenueFlow';
export { default as Efficiency } from './Efficiency';
export { default as TotalBalance } from './TotalBalance';
export { default as OverallBalance } from './OverallBalance';
export { default as UserCard } from './UserCard';
export { default as ItemSideBarSetting } from './ItemSideBarSetting';
export { default as FaqItem } from './Setting/Faq/FaqItem';
export { default as UpdateProfile } from './Profile';
export { default as AuthForm } from './AuthForm';
export { default as Modal } from './common/Modal';
export { default as Notification } from './common/Notification';
export { default as FallbackImage } from './FallbackImage';
export { default as ChatMember } from './ChatMember';
export { default as Conversation } from './Conversation';
export { default as BonusNotification } from './common/BonusNotification';
export { default as CardIssues } from './CustomerIssues/CardIssues';
export { default as CustomerIssues } from './CustomerIssues';
export { default as PinCode } from './common/PinCode';
export { default as Quill } from './Quill';
export { default as MessageAdmin } from './MessagesAdmin';
export { default as ListMessages } from './Conversation/ListMessage';
export { default as InputSendMessages } from './InputSendMessages';
export { default as PinCodeModal } from './PinCodeModal';
export { default as ProductsTable } from './ProductsTable';
export { default as UploadImages } from './UploadImages';
export { default as ProductForm } from './common/Table/Body/ProductForm';
export { default as RecentActivitiesTable } from './RecentActivitiesTable';
