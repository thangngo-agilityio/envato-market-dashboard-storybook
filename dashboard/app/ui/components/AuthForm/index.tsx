'use client';

import { ChangeEvent, memo, useCallback } from 'react';
import Link from 'next/link';
import { useRouter } from 'next/navigation';
import { ViewOffIcon, ViewIcon } from '@chakra-ui/icons';
import {
  Button,
  HStack,
  Text,
  VStack,
  Checkbox,
  useDisclosure,
  Box,
  Link as ChakraLink,
  Flex,
} from '@chakra-ui/react';

import { Controller, SubmitHandler } from 'react-hook-form';

// Hooks
import { useForm, useAuth } from '@/lib/hooks';

// Constants
import { ROUTES, AUTH_SCHEMA } from '@/lib/constants';

// Components
import { InputField } from '@/ui/components';

// Utils
import { isWindowDefined, validatePassword } from '@/lib/utils';

// Types
import { TUserDetail } from '@/lib/interfaces';
import { getMessaging } from 'firebase/messaging';

// Layouts
import { AuthFooter } from '@/ui/layouts';

// firebase
import { app, requestForToken } from '@/lib/utils';

type TAuthForm = Omit<TUserDetail, 'id' | 'createdAt'> & {
  confirmPassword: string;
  isAcceptPrivacyPolicy: boolean;
  isRemember: false;
};

interface AuthFormProps {
  isRegister?: boolean;
}

const AuthFormComponent = ({
  isRegister = false,
}: AuthFormProps): JSX.Element => {
  const { signIn, signUp } = useAuth();
  const router = useRouter();

  // Control form
  const {
    control,
    formState: {
      errors: { root },
      dirtyFields,
      defaultValues,
      isSubmitting,
    },
    handleSubmit,
    setError,
    clearErrors,
  } = useForm<TAuthForm>({
    defaultValues: {
      email: '',
      password: '',

      ...(!isRegister && {
        isRemember: false,
      }),
      ...(isRegister && {
        firstName: '',
        lastName: '',
        confirmPassword: '',
        isAcceptPrivacyPolicy: false,
      }),
    },
  });

  // Control show/hide password
  const { isOpen: isShowPassword, onToggle: onShowPassword } = useDisclosure();
  const { isOpen: isShowConfirmPassword, onToggle: onShowConfirmPassword } =
    useDisclosure();

  const messaging = isWindowDefined() ? getMessaging(app) : null;

  const isDisabledSubmitBtn: boolean = (() => {
    const getLength = (object: object): number => Object.keys(object).length;

    const isFillAllFields: boolean = isRegister
      ? getLength(dirtyFields) === getLength(defaultValues ?? {})
      : !!(dirtyFields.email && dirtyFields.password);

    return isSubmitting || !isFillAllFields;
  })();

  const renderPasswordIcon = useCallback(
    (isCorrect: boolean, callback: typeof onShowPassword): JSX.Element => {
      const Icon = isCorrect ? ViewIcon : ViewOffIcon;

      return (
        <Icon
          color="gray.400"
          w="25px"
          h="25px"
          cursor="pointer"
          onClick={callback}
        />
      );
    },
    [],
  );

  const handleLogin: SubmitHandler<TAuthForm> = useCallback(
    async (data) => {
      try {
        const { email, password, isRemember } = data;
        const fcmToken =
          (messaging && (await requestForToken(messaging))) || '';

        await signIn({ email, password, fcmToken }, isRemember);

        router.push(ROUTES.ROOT);
      } catch (error) {
        const { message } = error as Error;

        setError('root', { type: 'custom', message });
      }
    },
    [messaging, router, setError, signIn],
  );

  const handleRegister: SubmitHandler<TAuthForm> = useCallback(
    async (data) => {
      const {
        isAcceptPrivacyPolicy: _isAcceptPrivacyPolicy,
        confirmPassword: _confirmPassword,
        ...fieldValues
      } = data;
      try {
        const fcmToken =
          (messaging && (await requestForToken(messaging))) || '';

        await signUp({ ...fieldValues, fcmToken });

        router.push(ROUTES.ROOT);
      } catch (error) {
        const { message } = error as Error;

        setError('root', { type: 'custom', message });
      }
    },
    [messaging, router, setError, signUp],
  );

  const handleClearErrorMessage = useCallback(
    (
      field: keyof TAuthForm,
      isError: boolean,
      onChange: (value: string) => void,
    ) =>
      (data: string) => {
        if (isError) {
          clearErrors(field);
        }

        onChange(data);
      },
    [clearErrors],
  );

  const handleClearRootError = useCallback(
    () => clearErrors('root'),
    [clearErrors],
  );

  return (
    <Box
      id={!isRegister ? 'login-form' : 'register-form'}
      as="form"
      onSubmit={handleSubmit(!isRegister ? handleLogin : handleRegister)}
      w={{
        base: '100%',
        sm: 425,
        md: 460,
      }}
      margin="auto"
      pb={16}
      sx={{
        boxSizing: {
          base: 'border-box',
          md: 'unset',
        },
      }}
    >
      <VStack gap={6}>
        {isRegister && (
          <HStack
            gap={{
              base: 6,
              md: 10,
            }}
            w="100%"
            flexDirection={{
              base: 'column',
              md: 'row',
            }}
          >
            <Controller
              control={control}
              rules={AUTH_SCHEMA.FIRST_NAME}
              name="firstName"
              render={({ field, fieldState: { error } }) => (
                <InputField
                  variant="authentication"
                  placeholder="First name"
                  {...field}
                  isError={!!error}
                  errorMessages={error?.message}
                  isDisabled={isSubmitting}
                  onChange={handleClearErrorMessage(
                    'firstName',
                    !!error,
                    field.onChange,
                  )}
                  aria-label="first-name"
                />
              )}
            />
            <Controller
              control={control}
              rules={AUTH_SCHEMA.LAST_NAME}
              name="lastName"
              render={({ field, fieldState: { error } }) => (
                <InputField
                  variant="authentication"
                  placeholder="Last name"
                  {...field}
                  isError={!!error}
                  errorMessages={error?.message}
                  isDisabled={isSubmitting}
                  onChange={handleClearErrorMessage(
                    'lastName',
                    !!error,
                    field.onChange,
                  )}
                  aria-label="last-name"
                />
              )}
            />
          </HStack>
        )}
        <Controller
          rules={AUTH_SCHEMA.EMAIL}
          control={control}
          name="email"
          render={({ field: { value, onChange }, fieldState: { error } }) => {
            const handleChange = (valueInput: string) => {
              const sanitizedValue = valueInput.trim();

              onChange(sanitizedValue);
            };

            return (
              <InputField
                variant="authentication"
                placeholder="Email"
                isError={!!error?.message}
                errorMessages={error?.message}
                isDisabled={isSubmitting}
                value={value}
                onChange={handleChange}
                onBlur={handleClearRootError}
              />
            );
          }}
        />

        <Controller
          rules={{ validate: validatePassword }}
          control={control}
          name="password"
          render={({ field, fieldState: { error } }) => (
            <InputField
              type={isShowPassword ? 'text' : 'password'}
              variant="authentication"
              placeholder="Password"
              rightIcon={renderPasswordIcon(isShowPassword, onShowPassword)}
              isError={!!error?.message}
              errorMessages={error?.message}
              isDisabled={isSubmitting}
              {...field}
              onBlur={handleClearRootError}
            />
          )}
        />

        {/* Helpers */}
        {!isRegister && (
          <HStack justifyContent="space-between" w="100%" mt={6}>
            <Controller
              control={control}
              name="isRemember"
              render={({ field: { value, onChange } }) => (
                <Checkbox
                  aria-label="remember"
                  variant="round"
                  isChecked={value}
                  onChange={(e: ChangeEvent<HTMLInputElement>) =>
                    onChange(e.target.checked)
                  }
                  isDisabled={isSubmitting}
                  position="relative"
                >
                  <Text fontWeight="semibold">Remember me</Text>
                </Checkbox>
              )}
            />
            <Button
              as={Link}
              isDisabled={isSubmitting}
              href={`/${ROUTES.FORGOT_PASSWORD}`}
              _hover={{
                bg: 'transparent',
              }}
              bg="transparent"
              w="fit-content"
              aria-label="forgot password"
              color="text.currencyColor"
              fontWeight="semibold"
              textTransform="capitalize"
              textDecoration="underline"
              onClick={(e) => {
                if (isSubmitting) return e.preventDefault();
              }}
            >
              forgot password?
            </Button>
          </HStack>
        )}

        {isRegister && (
          <>
            <Controller
              control={control}
              rules={AUTH_SCHEMA.CONFIRM_PASSWORD}
              name="confirmPassword"
              render={({ field, fieldState: { error } }) => (
                <InputField
                  type={isShowConfirmPassword ? 'text' : 'password'}
                  variant="authentication"
                  placeholder="Confirm password"
                  rightIcon={renderPasswordIcon(
                    isShowConfirmPassword,
                    onShowConfirmPassword,
                  )}
                  {...field}
                  isError={!!error}
                  errorMessages={error?.message}
                  isDisabled={isSubmitting}
                  onChange={handleClearErrorMessage(
                    'confirmPassword',
                    !!error,
                    field.onChange,
                  )}
                />
              )}
            />

            <Flex gap={3}>
              <Controller
                control={control}
                rules={AUTH_SCHEMA.AGREE_POLICY}
                name="isAcceptPrivacyPolicy"
                render={({
                  field: { value, onChange },
                  fieldState: { error },
                }) => (
                  <Checkbox
                    size="md"
                    isChecked={value}
                    isDisabled={isSubmitting}
                    onChange={(e: ChangeEvent<HTMLInputElement>) =>
                      onChange(e.target.checked)
                    }
                    {...(error && {
                      sx: {
                        '> span': {
                          borderColor: 'danger.400',
                        },
                      },
                    })}
                  >
                    <Text
                      color="text.secondary"
                      fontSize={{ base: 'xs', md: 'md' }}
                      flex={1}
                      px={4}
                    >
                      By creating an account, you&apos;re agreeing to our {''}
                      <ChakraLink
                        href="#"
                        color="text.textDarkCheckbox"
                        cursor="pointer"
                      >
                        Privacy Policy
                      </ChakraLink>
                      , and {''}
                      <ChakraLink
                        href="#"
                        color="text.textDarkCheckbox"
                        cursor="pointer"
                      >
                        Electronics Communication Policy.
                      </ChakraLink>
                    </Text>
                  </Checkbox>
                )}
              />
            </Flex>
          </>
        )}
      </VStack>

      {/* Show API error */}
      <Box mb={7}>
        <Text color="red" textAlign="center" py={2} h={10}>
          {root?.message}
        </Text>
        <Button
          type="submit"
          role="button"
          aria-label={!isRegister ? 'Sign In' : 'Sign Up'}
          colorScheme="primary"
          bg="primary.300"
          textTransform="capitalize"
          form={!isRegister ? 'login-form' : 'register-form'}
          isDisabled={isDisabledSubmitBtn}
          isLoading={isSubmitting}
        >
          {!isRegister ? 'Sign In' : 'Sign Up'}
        </Button>
      </Box>

      <Flex justifyContent="center">
        <Text
          fontWeight="medium"
          textAlign="center"
          dangerouslySetInnerHTML={{
            __html: !isRegister
              ? 'Don&apos;t have an account?'
              : 'Already have an account?',
          }}
        />
        <Button
          as={Link}
          isDisabled={isSubmitting}
          href={!isRegister ? `/${ROUTES.REGISTER}` : `/${ROUTES.LOGIN}`}
          aria-label={!isRegister ? 'sign up' : 'sign in'}
          w="fit-content"
          py={0}
          _hover={{
            bg: 'transparent',
          }}
          color="text.primary"
          bg="transparent"
          fontWeight="semibold"
          textDecoration="underline"
          ml={2}
          onClick={(e) => {
            if (isSubmitting) return e.preventDefault();
          }}
        >
          {!isRegister ? 'Sign Up' : 'Sign In'}
        </Button>
      </Flex>
      <AuthFooter />
    </Box>
  );
};

const AuthForm = memo(AuthFormComponent);

export default AuthForm;
