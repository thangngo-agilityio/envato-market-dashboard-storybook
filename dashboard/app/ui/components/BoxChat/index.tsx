'use client';

import { memo, useCallback, useEffect, useRef, useState } from 'react';
import { Box, Heading, Flex } from '@chakra-ui/react';

// Components
import Message from './Message';

// Components
import { InputSendMessages } from '@/ui/components';

// Interface
import { TMessages } from '@/lib/interfaces';

// Stores
import { authStore } from '@/lib/stores';

// Hooks
import { getInfoRoomChat, useSubscribeToChat } from '@/lib/hooks';

// Firebase
import { doc, getDoc, setDoc } from 'firebase/firestore';
import { db } from '@/lib/utils';
import { AUTHENTICATION_ROLE, FIREBASE_CHAT } from '@/lib/constants';

const initialUserChat = {
  roomChatId: '',
  userId: '',
  adminId: '',
  avatarUrl: '',
  avatarAdminUrl: '',
  displayName: '',
};

const BoxChatComponent = () => {
  const [messages, setMessages] = useState<TMessages[]>([]);
  const user = authStore((state) => state.user);
  const hasPermission = user?.role === AUTHENTICATION_ROLE.MEMBER;
  const [userChat, setUserChat] = useState(initialUserChat);
  const boxRef = useRef<HTMLDivElement | null>(null);

  const fetchData = async () => {
    const usersData = await getInfoRoomChat(user);
    // Get user data

    // Check if usersData is undefined before accessing its properties
    if (usersData) {
      const res = await getDoc(
        doc(db, FIREBASE_CHAT.CHATS, usersData.roomChatId),
      );

      setUserChat(usersData);
      res.exists() ? setMessages(res.data().messages) : await createChatRoom();
    }
  };

  const createChatRoom = useCallback(async () => {
    // Get user data
    const usersData = await getInfoRoomChat(user);

    if (usersData) {
      await setDoc(doc(db, FIREBASE_CHAT.CHATS, usersData.roomChatId), {
        messages: [],
      });
    }
  }, [user]);

  useEffect(() => {
    fetchData();
  }, []);

  useSubscribeToChat(userChat.roomChatId, setMessages, boxRef);

  return (
    hasPermission && (
      <Box w="full" bg="background.body.quaternary" borderRadius="lg">
        <Flex
          direction="row"
          justifyContent="space-between"
          alignItems="center"
          justify="center"
          borderBottom="1px solid"
          borderBottomColor="border.tertiary"
          padding="24px 26px"
        >
          <Heading
            as="h3"
            fontWeight="semibold"
            color="text.primary"
            fontSize="2xl"
            textTransform="capitalize"
          >
            team chat
          </Heading>
        </Flex>

        <Box padding={{ base: '24px 20px', lg: '45px 35px' }}>
          <Box
            ref={boxRef}
            overflowX="auto"
            overflowY="scroll"
            css={{
              '&::-webkit-scrollbar': {
                width: 2,
              },
              '&::-webkit-scrollbar-track': {
                width: 2,
              },
              '&::-webkit-scrollbar-thumb': {
                background: 'gray',
                borderRadius: '24px',
              },
            }}
            maxHeight={361}
            padding={5}
          >
            {messages.map((message) => (
              <Message
                content={message.text}
                key={message.date.seconds}
                senderId={message.senderId}
                avatarAdmin={userChat.avatarAdminUrl}
                avatarUser={userChat.avatarUrl}
                superAdminId={userChat.adminId}
              />
            ))}
          </Box>
          <InputSendMessages boxRef={boxRef} />
        </Box>
      </Box>
    )
  );
};

const BoxChat = memo(BoxChatComponent);

export default BoxChat;
