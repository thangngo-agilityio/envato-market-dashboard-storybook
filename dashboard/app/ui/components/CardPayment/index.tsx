'use client';

import { FormEvent, memo, useCallback } from 'react';
import { Box, Heading, useDisclosure, useToast } from '@chakra-ui/react';
import { SubmitHandler } from 'react-hook-form';
import { AxiosError } from 'axios';

// Hooks
import {
  useAuth,
  useForm,
  useGetUserDetails,
  useMoney,
  usePinCode,
  useWallet,
} from '@/lib/hooks';

// Components
import { PinCodeModal } from '@/ui/components';
import CardBalance from './CardBalance';
import UserSelector from './UserSelector';
import EnterMoney from './EnterMoney';

// Stores
import { authStore } from '@/lib/stores';

// Types
import { TPinCodeForm, TSendMoney } from '@/lib/interfaces';

// Constants
import { ERROR_MESSAGES, STATUS, SUCCESS_MESSAGES } from '@/lib/constants';

// Utils
import { customToast, getErrorMessageFromAxiosError } from '@/lib/utils';

// services
import { TMoneyResponse } from '@/lib/services';

export type TTransfer = {
  amount: string;
  memberId: string;
  userId: string;
};

const CardPaymentComponent = (): JSX.Element => {
  const user = authStore((state) => state.user);

  const { setUser } = useAuth();

  const toast = useToast();

  const {
    control,
    handleSubmit: handleSubmitSendMoney,
    formState: { isValid, isSubmitting },
    reset: resetSendMoneyForm,
  } = useForm<TTransfer>({
    defaultValues: {
      memberId: '',
      amount: '',
      userId: user?.id,
    },
  });

  const { currentWalletMoney } = useWallet(user?.id);

  const { filterDataUser } = useGetUserDetails(user?.id || '');

  const {
    isOpen: isSetPinCodeModalOpen,
    onClose: onCloseSetPinCodeModal,
    onOpen: onOpenSetPinCodeModal,
  } = useDisclosure();

  const {
    isOpen: isConfirmPinCodeModalOpen,
    onClose: onCloseConfirmPinCodeModal,
    onOpen: onOpenConfirmPinCodeModal,
  } = useDisclosure();

  const { handleSetPinCode, handleConfirmPinCode } = usePinCode();

  const {
    control: setPinCodeControl,
    handleSubmit: handleSubmitSetPinCode,
    formState: { isValid: isSetValid, isSubmitting: isSetSubmitting },
    reset: resetSetPinCodeForm,
  } = useForm<TPinCodeForm>({});

  const {
    control: confirmPinCodeControl,
    handleSubmit: handleSubmitConfirmPinCode,
    formState: { isValid: isConfirmValid, isSubmitting: isConfirmSubmitting },
    reset: resetConfirmPinCodeForm,
  } = useForm<TPinCodeForm>({
    defaultValues: {
      pinCode: '',
    },
    mode: 'onSubmit',
    reValidateMode: 'onSubmit',
  });

  const { sendMoneyToUserWallet } = useMoney();

  const getMemberId = useCallback(
    (email: string): string =>
      filterDataUser?.find(
        (user) =>
          user.email.trim().toLocaleLowerCase() === email.trim().toLowerCase(),
      )?._id || '',
    [filterDataUser],
  );

  const hasPinCode = user?.pinCode;

  const handleTransferMoneySuccess = (success: {
    title: string;
    description: string;
  }) => {
    toast(customToast(success.title, success.description, STATUS.SUCCESS));
    if (user?.bonusTimes) {
      setUser({
        user: {
          ...user,
          bonusTimes: --user.bonusTimes,
        },
      });
    }
  };

  const handleTransferMoneyError = (
    error: Error,
    defaultError: {
      title: string;
      description: string;
    },
  ) => {
    const responseErrorMessage = getErrorMessageFromAxiosError(
      error as AxiosError<TMoneyResponse>,
      defaultError.description,
    );

    toast(customToast(defaultError.title, responseErrorMessage, STATUS.ERROR));
  };

  const handleOnSubmitSendMoney = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    if (!hasPinCode) {
      onOpenSetPinCodeModal();
    } else {
      onOpenConfirmPinCodeModal();
    }
  };

  const onSubmitSendMoney: SubmitHandler<TTransfer> = useCallback(
    (data) => {
      const submitData: TSendMoney = {
        ...data,
        memberId: getMemberId(data.memberId),
        amount: Number(data.amount),
      };

      sendMoneyToUserWallet(submitData, {
        onSuccess: () =>
          handleTransferMoneySuccess(SUCCESS_MESSAGES.SEND_MONEY),
        onError: (error) =>
          handleTransferMoneyError(error, ERROR_MESSAGES.ADD_MONEY),
      });
      resetSendMoneyForm();
    },
    [getMemberId, resetSendMoneyForm, sendMoneyToUserWallet],
  );

  const onSubmitPinCode: SubmitHandler<TPinCodeForm> = useCallback(
    async (data) => {
      if (user) {
        data.userId = user.id;
        if (!hasPinCode) {
          try {
            await handleSetPinCode(data);

            setUser({ user: { ...user, pinCode: data.pinCode } });

            onCloseSetPinCodeModal();

            resetSetPinCodeForm();

            toast(
              customToast(
                SUCCESS_MESSAGES.SET_PIN_CODE.title,
                SUCCESS_MESSAGES.SET_PIN_CODE.description,
                STATUS.SUCCESS,
              ),
            );
          } catch (error) {
            toast(
              customToast(
                ERROR_MESSAGES.SET_PIN_CODE.title,
                ERROR_MESSAGES.SET_PIN_CODE.description,
                STATUS.ERROR,
              ),
            );
          }
        } else {
          try {
            await handleConfirmPinCode(data);
            onCloseConfirmPinCodeModal();
            resetConfirmPinCodeForm({
              pinCode: '',
            });

            await handleSubmitSendMoney(onSubmitSendMoney)();
            resetSendMoneyForm();

            toast(
              customToast(
                SUCCESS_MESSAGES.CONFIRM_PIN_CODE.title,
                SUCCESS_MESSAGES.CONFIRM_PIN_CODE.description,
                STATUS.SUCCESS,
              ),
            );
          } catch (error) {
            toast(
              customToast(
                ERROR_MESSAGES.CONFIRM_PIN_CODE.title,
                ERROR_MESSAGES.CONFIRM_PIN_CODE.description,
                STATUS.ERROR,
              ),
            );
            resetConfirmPinCodeForm();
          }
        }
      }
    },
    [
      handleConfirmPinCode,
      handleSetPinCode,
      handleSubmitSendMoney,
      hasPinCode,
      onCloseConfirmPinCodeModal,
      onCloseSetPinCodeModal,
      onSubmitSendMoney,
      resetConfirmPinCodeForm,
      resetSendMoneyForm,
      resetSetPinCodeForm,
      setUser,
      toast,
      user,
    ],
  );

  const handleCloseSetPinCodeModal = useCallback(() => {
    onCloseSetPinCodeModal();
    resetSetPinCodeForm();
  }, [onCloseSetPinCodeModal, resetSetPinCodeForm]);

  const handleCloseConfirmPinCodeModal = useCallback(() => {
    onCloseConfirmPinCodeModal();
    resetConfirmPinCodeForm();
  }, [onCloseConfirmPinCodeModal, resetConfirmPinCodeForm]);

  return (
    <>
      <Box
        p={4}
        w="full"
        bg="background.body.quaternary"
        py={{ base: 4, md: 5 }}
        px={{ base: 4, md: 10 }}
        borderRadius="lg"
      >
        <Heading
          as="h3"
          fontWeight="bold"
          color="text.primary"
          fontSize="lg"
          mb={3}
          textTransform="capitalize"
        >
          my wallet
        </Heading>

        <CardBalance balance={currentWalletMoney?.balance || 0} />

        <Box as="form" mt={4} onSubmit={handleOnSubmitSendMoney}>
          <UserSelector control={control} listUser={filterDataUser} />
          <EnterMoney isDisabled={!isValid || isSubmitting} control={control} />
        </Box>
      </Box>
      {/*Set/Confirm PIN code Modal */}
      <PinCodeModal
        title={
          isSetPinCodeModalOpen
            ? 'Please set the PIN code to your account'
            : 'Please enter your PIN code'
        }
        control={hasPinCode ? confirmPinCodeControl : setPinCodeControl}
        isOpen={isSetPinCodeModalOpen || isConfirmPinCodeModalOpen}
        isDisabled={
          hasPinCode
            ? !isConfirmValid || isConfirmSubmitting
            : !isSetValid || isSetSubmitting
        }
        onclose={
          isSetPinCodeModalOpen
            ? handleCloseSetPinCodeModal
            : handleCloseConfirmPinCodeModal
        }
        onSubmit={
          hasPinCode
            ? handleSubmitConfirmPinCode(onSubmitPinCode)
            : handleSubmitSetPinCode(onSubmitPinCode)
        }
      />
    </>
  );
};

const CardPayment = memo(CardPaymentComponent);

export default CardPayment;
