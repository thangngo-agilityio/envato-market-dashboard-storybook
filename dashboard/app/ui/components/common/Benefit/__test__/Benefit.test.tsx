import '@testing-library/jest-dom';

// component
import { Benefit } from '@/ui/components';

// Constants
import { ROUTES } from '@/lib/constants';

const { render } = testLibReactUtils;

describe('Benefit render', () => {
  test('renders Benefit with pathName is Login', () => {
    const { container } = render(<Benefit pathName={`/${ROUTES.LOGIN}`} />);

    expect(container).toMatchSnapshot();
  });

  test('renders Benefit with pathName is SignUp', () => {
    const { container } = render(<Benefit pathName={`/${ROUTES.REGISTER}`} />);

    expect(container).toMatchSnapshot();
  });
});
