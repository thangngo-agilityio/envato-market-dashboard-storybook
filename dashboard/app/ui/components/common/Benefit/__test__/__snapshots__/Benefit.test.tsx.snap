// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Benefit render renders Benefit with pathName is Login 1`] = `
<div>
  <div
    class="css-z86n61"
  >
    <img
      alt="This is sign in image"
      data-nimg="1"
      decoding="async"
      height="596"
      loading="lazy"
      src="/icons/signin.svg"
      style="color: transparent; background-size: cover; background-position: 50% 50%; background-repeat: no-repeat;"
      width="699"
    />
    <div
      class="css-1saybp7"
    >
      <img
        alt="This is square image"
        data-nimg="fill"
        decoding="async"
        loading="lazy"
        src="/icons/square.svg"
        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent; background-size: cover; background-position: 50% 50%; background-repeat: no-repeat;"
      />
    </div>
    <div
      class="css-1uw9kux"
    >
      <img
        alt="This is vline image"
        data-nimg="fill"
        decoding="async"
        loading="lazy"
        src="/icons/vline.svg"
        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent; background-size: cover; background-position: 50% 50%; background-repeat: no-repeat;"
      />
    </div>
    <div
      class="css-he7guj"
    >
      <img
        alt="This is dotted image"
        data-nimg="fill"
        decoding="async"
        loading="lazy"
        src="/icons/dotted.svg"
        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent; background-size: cover; background-position: 50% 50%; background-repeat: no-repeat;"
      />
    </div>
    <div
      class="css-18anm8l"
    >
      <h2
        class="chakra-heading css-1qw8f9f"
      >
        Speady, Easy and Fast
      </h2>
      <p
        class="chakra-text css-1omtyor"
      >
        BankCo. help you set saving goals, earn cash back offers, Go to disclaimer for more details and get paychecks up to two days early. Get a
        <span
          class="chakra-text css-xety8z"
        >
          $20
        </span>
        bonus when you receive qualifying direct deposits
      </p>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;

exports[`Benefit render renders Benefit with pathName is SignUp 1`] = `
<div>
  <div
    class="css-z86n61"
  >
    <img
      alt="This is sign up image"
      data-nimg="1"
      decoding="async"
      height="480"
      loading="lazy"
      src="/icons/signup.svg"
      style="color: transparent; background-size: cover; background-position: 50% 50%; background-repeat: no-repeat;"
      width="699"
    />
    <div
      class="css-1saybp7"
    >
      <img
        alt="This is square image"
        data-nimg="fill"
        decoding="async"
        loading="lazy"
        src="/icons/square.svg"
        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent; background-size: cover; background-position: 50% 50%; background-repeat: no-repeat;"
      />
    </div>
    <div
      class="css-1uw9kux"
    >
      <img
        alt="This is vline image"
        data-nimg="fill"
        decoding="async"
        loading="lazy"
        src="/icons/vline.svg"
        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent; background-size: cover; background-position: 50% 50%; background-repeat: no-repeat;"
      />
    </div>
    <div
      class="css-he7guj"
    >
      <img
        alt="This is dotted image"
        data-nimg="fill"
        decoding="async"
        loading="lazy"
        src="/icons/dotted.svg"
        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent; background-size: cover; background-position: 50% 50%; background-repeat: no-repeat;"
      />
    </div>
    <div
      class="css-18anm8l"
    >
      <h2
        class="chakra-heading css-1qw8f9f"
      >
        Speady, Easy and Fast
      </h2>
      <p
        class="chakra-text css-1omtyor"
      >
        BankCo. help you set saving goals, earn cash back offers, Go to disclaimer for more details and get paychecks up to two days early. Get a
        <span
          class="chakra-text css-xety8z"
        >
          $20
        </span>
        bonus when you receive qualifying direct deposits
      </p>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
